### PR TITLE
[Feature] Refactor translation models

### PR DESCRIPTION
**Pull request by plyfager**
Add a base module for the translation models.

### CODE REVIEW

**Comment by nbei** [comment]
You may fix the linting error when available.

**Comment by plyfager** [comment]
There are mainly two changes in this refactorization.
The first point is that we abandon `forward(img_a, img_b, ...)` in the forward function and instead pass in one image and the style like `forward(img, style, ...)`. In `demo/translation_demo.py`, you can decide the style by using `--style {STYLE}`. For the second point, we group different loss items into the three types in `BaseGAN` and use the same training process.

There are still some items unfinished:

- [ ] Configs and checkpoints
- [ ] Evaluation during training and inference 
- [ ] Docstrings
- [ ] Documents

File: mmgen/models/gans/base_translation_model.py [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+import torch.nn as nn
+
+from ..builder import MODELS, build_module
+from .base_gan import BaseGAN
+
+
+@MODELS.register_module()
+class BaseTranslationModel(BaseGAN):
+    """Basic translation Model.
+
```
Comment (nbei) [inline mmgen/models/gans/base_translation_model.py:12]: You may specify the basic design logic for this base model in the docstring.

File: mmgen/models/gans/base_translation_model.py [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+    Args:
+        generator (dict): Config for the generator.
+        discriminator (dict): Config for the discriminator.
+        gan_loss (dict): Config for the gan loss.
+        default_style (str): Default output style.
+        reachable_styles (list[str], optional): Styles that can be generated by
+            the model. Defaults to [].
```
Comment (nbei) [inline mmgen/models/gans/base_translation_model.py:19]: Typically, we will not use `[]` or `{}` as the default argument. Can we use `None` as the default value?

File: mmgen/models/gans/base_translation_model.py [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+        related_styles (list[str], optional): Styles involved in training and
```
Comment (nbei) [inline mmgen/models/gans/base_translation_model.py:20]: In the docstring, you may tell users how to set `related_styles`, `default_style`, and `reachable_styles`. In addition, are different algorithms using different definitions for these arguments?

File: mmgen/models/gans/base_translation_model.py [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+        super().__init__()
+        self.fp16_enabled = False
+        self._default_style = default_style
+        self._reachable_styles = reachable_styles
+        self._related_styles = related_styles
+        assert self._default_style in self._reachable_styles
+        assert set(self._reachable_styles) <= set(self._related_styles)
+
+        # TODO: Building G and D in following way can be used for
+        # Pix2Pix and CycleGAN
```
Comment (nbei) [inline mmgen/models/gans/base_translation_model.py:55]: Has it finished?

File: mmgen/models/gans/base_translation_model.py [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+
+        self._gen_cfg = deepcopy(generator)
+        # build style generators
+        self.generators = nn.ModuleDict()
+        for style in self._reachable_styles:
+            self.generators[style] = build_module(generator)
+
+        self._disc_cfg = deepcopy(discriminator)
+        # build style-aware discriminators
```
Comment (nbei) [inline mmgen/models/gans/base_translation_model.py:64]: `style-aware` is not accurate. You may use `category-aware` or other names.

File: mmgen/models/gans/base_translation_model.py [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+            self._parse_test_cfg()
+
+    def _parse_train_cfg(self):
+        """Parsing train config and set some attributes for training."""
+        if self.train_cfg is None:
+            self.train_cfg = dict()
+        # control the work flow in train step
+        self.disc_steps = self.train_cfg.get('disc_steps', 1)
+
+        self.disc_init_steps = (0 if self.train_cfg is None else
```
Comment (nbei) [inline mmgen/models/gans/base_translation_model.py:109]: `self.train_cfg` cannot be `None`.

File: mmgen/models/gans/base_translation_model.py [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+                                self.train_cfg.get('disc_init_steps', 0))
+
+        self.real_img_key = self.train_cfg.get('real_img_key', 'real_img')
```
Comment (nbei) [inline mmgen/models/gans/base_translation_model.py:112]: Do you still need it?

File: mmgen/models/gans/base_translation_model.py [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+            img (tensor): Input image tensor.
+            test_mode (bool): Whether in test mode or not. Default: False.
+            kwargs (dict): Other arguments.
+        """
+        if not test_mode:
+            return self.forward_train(img, **kwargs)
+
+        return self.forward_test(img, **kwargs)
+
+    def forward_train(self, img, style, **kwargs):
```
Comment (nbei) [inline mmgen/models/gans/base_translation_model.py:137]: You may be careful to use `style`. Style transfer and image translation have obvious differences. Thus, you may consider replacing it with other words, like 'target'.

File: mmgen/models/gans/base_translation_model.py [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+        Args:
+            img (tensor): Input image tensor.
+            test_mode (bool): Whether in test mode or not. Default: False.
+            kwargs (dict): Other arguments.
+
+        Returns:
+            Dict: Forward results.
+        """
+        # This is a trick for Pix2Pix and cyclegan
+        self.train()
```
Comment (nbei) [inline mmgen/models/gans/base_translation_model.py:164]: If this is a trick for Pix2Pix and CycleGAN, it should be placed in those models instead of here.

File: mmgen/models/gans/base_translation_model.py [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+        styled_img = self.translation(img, style=style)
+        results = dict(orig_img=img.cpu(), styled_img=styled_img.cpu())
+        return results
+
+    def is_style_reachable(self, style):
```
Comment (nbei) [inline mmgen/models/gans/base_translation_model.py:169]: Be careful about `style`. In the Pix2Pix or CycleGAN, do they use this word?

File: mmgen/models/gans/cyclegan.py [hunk 11]
```diff
@@ -23,95 +18,55 @@ class CycleGAN(BaseGAN):
...
+            generator,
+            discriminator,
+            gan_loss,
+            default_style,
+            reachable_styles=reachable_styles,
+            related_styles=related_styles,
+            disc_auxiliary_loss=disc_auxiliary_loss,
+            gen_auxiliary_loss=gen_auxiliary_loss,
+            train_cfg=train_cfg,
+            test_cfg=test_cfg)
```
Comment (nbei) [inline mmgen/models/gans/cyclegan.py:63]: If you have inherited `BaseTranslationModel`, you can write as follows:

```python
class CycleGAN(BaseTranslationModel):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)

```

File: mmgen/models/gans/cyclegan.py [hunk 12]
```diff
@@ -145,322 +99,168 @@ def get_module(self, module):
...
+                fake_pred, target_is_real=True, is_disc=False)
+
+        # gen auxiliary loss
+        if self.with_gen_auxiliary_loss:
+            for loss_module in self.gen_auxiliary_losses:
+                loss_ = loss_module(outputs)
+                if loss_ is None:
+                    continue
+
+                # mmcv.print_log(f'get loss for {loss_module.name()}')
```
Comment (nbei) [inline mmgen/models/gans/cyclegan.py:164]: remove dirty comments

File: mmgen/models/gans/base_translation_model.py (hunk 3 above)
Comment (LeoXing1996) [reply mmgen/models/gans/base_translation_model.py:20]: Maybe we should use 'domain'?

**Review by LeoXing1996 — COMMENTED** [review]
1. In my option, `base_translation_model` should not inherit `BaseGAN` . It should contain *a group of GANs* or a group of other generative models. Otherwise, the translation would be constrained in GANs.
2. It seems translation methods contain two-part of loss: within-domain loss and cross-domain loss. We can define two functions called `_get_within_domain_loss` and `_get_cross_domain_loss`, and implement them in child.  In `_get_within_domain_loss` loss can be calculated by generative models and in `_cross_domain_loss` would be calculated by `TranslationModel`.
3. To prepare data used for loss calculation, we can define a function called `translation_image` , then we can perform translation behavior such as `A->B`, `A->B->A`  under the guidance of passed arguments. Then in `forward_train`, we may perform all translation need to calculate losses. And in `forward_test`, arguments can be directly passed to `translation_image`, users can get desired images like use `sample_from_noise` in GAN-models.

The above suggestions are just trying to compatible all kinds of translation models.
I think we should survey the design of modern translation models with complex update strategies such as Star-GAN (v2) and multi-model generation methods.

File: mmgen/models/gans/base_translation_model.py (hunk 5 above)
Comment (plyfager) [reply mmgen/models/gans/base_translation_model.py:64]: Is it fine to use `domain discriminators `?

Comment (LeoXing1996) [reply mmgen/models/gans/base_translation_model.py:64]: I think it's not necessary.
For a GAN-based translation model, discrimintor and generator are all domain-specificed. Otherwise, those modules  would builded by `build_disc/gen_auxiliary`.

File: mmgen/models/gans/cyclegan.py [hunk 13]
```diff
@@ -145,322 +99,168 @@ def get_module(self, module):
...
         # forward generators
-        outputs = self.forward(img_a, img_b, meta, test_mode=False)
+        outputs = dict()
+        for style in self._reachable_styles:
+            # fetch data by style
+            another_style = self._get_opposite_style(style)
+            img = data_batch[f'img_{another_style}']
+            # stylized process
+            results = self.forward(img, test_mode=False, style=style)
+            outputs[f'src_{another_style}'] = results['orig_img']
```
[the remark this answers is not in the document]
Comment (plyfager) [reply mmgen/models/gans/cyclegan.py:221]: I just use 'real_{domain}', 'fake_{domain}', 'source', 'target' as variable names.

File: mmgen/models/gans/cyclegan.py [hunk 14]
```diff
@@ -1,18 +1,13 @@
...
-from mmgen.models.misc import tensor2img
+from mmgen.models.builder import MODELS
 from ..common import GANImageBuffer, set_requires_grad
-from .base_gan import BaseGAN
+from .base_translation_model import BaseTranslationModel
 
 
 @MODELS.register_module()
-class CycleGAN(BaseGAN):
+class CycleGAN(BaseTranslationModel):
```
[the remark this answers is not in the document]
Comment (plyfager) [reply mmgen/models/gans/cyclegan.py:10]: I would create a class `StaticTranslationGAN` inherited from `BaseTranslationModel` and `BaseGAN` to support `Pix2Pix` and `CycleGAN`.

File: mmgen/models/gans/base_translation_model.py [hunk 15]
```diff
@@ -0,0 +1,200 @@
+from copy import deepcopy
```
Comment (LeoXing1996) [inline mmgen/models/gans/base_translation_model.py:1]: Please don't forget to add LICENSE at the start of the file.

File: mmgen/models/gans/base_translation_model.py [hunk 16]
```diff
@@ -0,0 +1,135 @@
+from abc import ABCMeta, abstractmethod
+from copy import deepcopy
+
+import torch.nn as nn
+
+from ..builder import MODELS
+
+
+@MODELS.register_module()
+class BaseTranslationModel(nn.Module, metaclass=ABCMeta):
```
Comment (LeoXing1996) [inline mmgen/models/gans/base_translation_model.py:10]: This design seems nice to me. However, this file should under a separate folder, such as `mmgen/models/translation_models/base_translation_model.py`.
(Maybe github has bugs on name and path display after using `git mv`.)

File: mmgen/models/gans/base_translation_model.py [hunk 17]
```diff
@@ -0,0 +1,135 @@
...
+    def is_domain_reachable(self, domain):
+        """Whether image of this domain can be generated."""
+        return domain in self._reachable_domains
+
+    def _get_other_domains(self, domain):
+        """get other domains."""
+        return list(set(self._related_domains) - set([domain]))
+
+    @abstractmethod
+    def _get_domain_generator(self, domain):
```
[the remark this answers is not in the document]
Comment (plyfager) [reply mmgen/models/gans/base_translation_model.py:117]: `_get_domain_generator` returns a generator that can generate images in a specific domain. Sure, `_get_target_generator` sounds more reasonable.

File: mmgen/datasets/unpaired_image_dataset.py [hunk 18]
```diff
@@ -29,9 +29,16 @@ class UnpairedImageDataset(Dataset):
...
+        domain_a (str, optional): Name of domain A. Defaults to 'img_a'.
+        domain_b (str, optional): Name of domain B. Defaults to 'img_b'.
     """
 
-    def __init__(self, dataroot, pipeline, test_mode=False):
+    def __init__(self,
+                 dataroot,
+                 pipeline,
+                 test_mode=False,
+                 domain_a='a',
```
[the remark this answers is not in the document]
Comment (plyfager) [reply mmgen/datasets/unpaired_image_dataset.py:40]: I would set the default value to None and assert these arguments are strings.

File: mmgen/datasets/pipelines/loading.py [hunk 19]
```diff
@@ -95,6 +95,8 @@ class LoadPairedImageFromFile(LoadImageFromFile):
     Args:
         io_backend (str): io backend where images are store. Default: 'disk'.
         key (str): Keys in results to find corresponding path. Default: 'gt'.
+        domain_a (str, optional): Style of domain A. Defaults to 'img_a'.
```
[the remark this answers is not in the document]
Comment (plyfager) [reply mmgen/datasets/pipelines/loading.py:98]: I would set the default value to None and assert these arguments are strings. The exact meaning of `domain_a` and `domain_b` will be specified.

File: mmgen/models/gans/cyclegan.py [hunk 20]
```diff
@@ -146,322 +61,169 @@ def get_module(self, module):
...
 
         # forward generators
-        outputs = self.forward(img_a, img_b, meta, test_mode=False)
+        outputs = dict()
+        for target_domain in self._reachable_domains:
+            # fetch data by domain
+            source_domain = self.get_other_domains(target_domain)[0]
+            img = data_batch[f'img_{source_domain}']
+            # translation process
+            results = self.forward(
```
Comment (nbei) [inline mmgen/models/gans/cyclegan.py:180]: Since this is an instance of `nn.Module`, you would better use `self(xx)` instead of `self.forward()`.

File: mmgen/models/gans/cyclegan.py [hunk 21]
```diff
@@ -146,322 +61,169 @@ def get_module(self, module):
...
+                img, test_mode=False, target_domain=target_domain)
+            outputs[f'real_{source_domain}'] = results['source']
+            outputs[f'fake_{target_domain}'] = results['target']
+            # cycle process
+            results = self.forward(
```
Comment (nbei) [inline mmgen/models/gans/cyclegan.py:185]: Same problem as aforementioned.

File: mmgen/models/gans/pix2pix.py [hunk 22]
```diff
@@ -1,340 +1,181 @@
...
-import numpy as np
 import torch
+from torch.nn.parallel.distributed import _find_tensors
 
-from mmgen.models.builder import MODELS, build_module
-from mmgen.models.misc import tensor2img
+from mmgen.models.builder import MODELS
 from ..common import set_requires_grad
-from .base_gan import BaseGAN
+from .static_translation_gan import StaticTranslationGAN
```
Comment (nbei) [inline mmgen/models/gans/pix2pix.py:8]: You can create a new folder `mmgen/models/translation_models` and tidy related files.

File: mmgen/models/gans/pix2pix.py [hunk 23]
```diff
@@ -1,340 +1,181 @@
...
-            save_path (str, optional): If given a valid str path, the results
-                will be saved in this path. Default: None.
-            iteration (int, optional): Iteration number. Default: None.
+            img (tensor): Input image tensor.
+            target_domain (str): Target domain of output image.
+            kwargs (dict): Other arguments.
 
         Returns:
-            dict: Dict of forward and evaluation results for testing.
+            Dict: Forward results.
```
Comment (nbei) [inline mmgen/models/gans/pix2pix.py:32]: Dict ---> dict

File: mmgen/models/gans/pix2pix.py [hunk 24]
```diff
@@ -1,340 +1,181 @@
...
+            # dirty walkround for not providing running status
+            if not hasattr(self, 'iteration'):
+                self.iteration = 0
+            curr_iter = self.iteration
 
         # forward generator
-        outputs = self.forward(img_a, img_b, meta, test_mode=False)
-
+        outputs = dict()
+        results = self.forward(
```
Comment (nbei) [inline mmgen/models/gans/pix2pix.py:128]: Same problem as aforementioned.

File: configs/_base_/models/pix2pix_vanilla_unet_bn.py [hunk 25]
```diff
@@ -23,7 +25,15 @@
         real_label_val=1.0,
         fake_label_val=0.0,
         loss_weight=1.0),
-    pixel_loss=dict(type='L1Loss', loss_weight=100.0, reduction='mean'))
+    default_domain=target_domain,
```
[the remark this answers is not in the document]
Comment (plyfager) [reply configs/_base_/models/pix2pix_vanilla_unet_bn.py:28]: I suggest we categorize these configs in a new PR.
